compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPUDistributeContract.cpp

```diff
@@ -0,0 +1,269 @@
+      // accumulate to the same vector.
+      for (int k = 0; k < kBatch; ++k) {
+        if (!getOperandBatchOffsets(contractKind, k, resultBatchOffsets,
+                                    resultLayout, lhsBatchOffsets,
+                                    rhsBatchOffsets, lhsLayout, rhsLayout)) {
+          return rewriter.notifyMatchFailure(
+              contractOp, "cannot deduce lhs/rhs batch offsets");
+        }
+        LLVM_DEBUG({
+          llvm::dbgs() << "current lhs batch offsets: [";
+          llvm::interleaveComma(lhsBatchOffsets, llvm::dbgs());
+          llvm::dbgs() << "]\n";
+          llvm::dbgs() << "current rhs batch offsets: [";
+          llvm::interleaveComma(rhsBatchOffsets, llvm::dbgs());
+          llvm::dbgs() << "]\n";
+        });
+        Value lhsSlice = rewriter.create<vector::ExtractOp>(
+            loc, getDistributed(rewriter, contractOp.getLhs(), lhsLayout),
+            lhsBatchOffsets);
+        Value rhsSlice = rewriter.create<vector::ExtractOp>(
+            loc, getDistributed(rewriter, contractOp.getRhs(), rhsLayout),
+            rhsBatchOffsets);
+        accSlice = computeMMA(rewriter, loc, lhsSlice, rhsSlice, accSlice,
+                              aVectorType, bVectorType, cVectorType,
+                              mfmaAttr.getIntrinsic().getValue());
+      }
+      finalTile = rewriter.create<vector::InsertOp>(loc, accSlice, finalTile,
+                                                    resultBatchOffsets);
+    }
+
+    replaceOpWithDistributedValues(rewriter, contractOp, finalTile);
+    return success();
+  }
+
+  // Gets the batch size for matmul K dimensions.
+  std::optional<int64_t> getKBatchSize(ContractKind kind,
+                                       NestedLayoutAttr lhsLayout,
+                                       NestedLayoutAttr rhsLayout) const {
+    int64_t lhsKBatch = 0, rhsKBatch = 0;
+    if (kind == ContractKind::MK_KN_MN) {
+      lhsKBatch = lhsLayout.getBatchesPerSubgroup()[1];
+      rhsKBatch = rhsLayout.getBatchesPerSubgroup()[0];
+    } else {
+      return std::nullopt;
+    }
+
+    if (lhsKBatch != rhsKBatch)
+      return std::nullopt;
+    return lhsKBatch;
+  }
+
+  // Given a contract op's batch |resultOffsets|, gets its batch offsets for
+  // both LHS and RHS.
+  bool getOperandBatchOffsets(
+      ContractKind kind, int64_t kOffset,
+      // Copy intentionally given we need to mutate in the function body
+      SmallVector<int64_t, 2> resultOffsets, NestedLayoutAttr resultLayout,
+      SmallVector<int64_t, 2> &lhsOffsets, SmallVector<int64_t, 2> &rhsOffsets,
+      NestedLayoutAttr lhsLayout, NestedLayoutAttr rhsLayout) const {
+    // The result offsets are permutated and we need to revert the permutation.
+    // The following works based on the fact that for 2-D cases, a permuation
+    // vector's reverse is just itself. So this only works for 2-D cases.
+    applyPermutationToVector(resultOffsets, resultLayout.getBatchOrder());
```

Comment:
You can use `invertPermutationVector(resultLayout.getBatchOrder())` (or just `invertPermutation`, I forget exactly), although I think the offsets currently might not be permuted per my comment above.

Reply:
no need of this anymore given I can pass in the original unpermutated batch offsets.